### PR TITLE
Fix mutex initialization via PTHREAD_MUTEX_INITIALIZER

### PR DESCRIPTION
**Pull request by lucasccordeiro**
Our pthread operational model was not considering initialization of mutexes via PTHREAD_MUTEX_INITIALIZER. As a result, we used to produce incorrect results as described by this PR https://github.com/esbmc/esbmc/issues/548. Here, for each variant of `pthread_mutex_lock`, we check whether the mutex has been initialized via PTHREAD_MUTEX_INITIALIZER. If so, we call `pthread_mutex_init` similar to the original implementation of the `pthread_mutex_lock` (see https://opensource.apple.com/source/Libc/Libc-167/pthreads.subproj/pthread_mutex.c).

### CODE REVIEW

**Comment by kunjsong01** [comment]
LGTM.

This patch equates the semantics of `PTHREAD_MUTEX_INITIALIZER` macro to `pthread_mutex_init` function. The test cases in this patch aim to test the true branch of the condition `(mutex->__align == _PTHREAD_MUTEX_SIG_init)`, i.e. when using the initializer macro.  

The false branch of this condition is already covered by the following test cases in our regression test suite: 
```
../regression/esbmc-unix2/01_pthread10/main.c:16:  pthread_mutex_init(&mutex, NULL);
../regression/esbmc-unix2/01_pthread11/main.c:23:   pthread_mutex_init( &mx1, NULL );
../regression/esbmc-unix2/01_pthread11/main.c:31:   pthread_mutex_init( &mx2, NULL );
../regression/esbmc-unix2/01_pthread12/main.c:24:   r = pthread_mutex_init( &mx1, &attr );
../regression/esbmc-unix2/01_pthread13/main.c:31:  pthread_mutex_init(&m1, NULL);
...
...
```

At the time of writing this comment, the regressions are still running. I think it's safe to merge when the above test cases also pass, which gvies us the confidence that the new function `pthread_mutex_initializer` does not break the existing flow w.r.t `pthread_mutex_init` function.

File: src/c2goto/library/pthread_lib.c [hunk 1]
```diff
@@ -291,10 +295,25 @@ __ESBMC_HIDE:;
   return 0;
 }
 
+int pthread_mutex_initializer(pthread_mutex_t *mutex)
+{
+  // check whether this mutex has been initialized via
+  // PTHREAD_MUTEX_INITIALIZER
+#ifdef __APPLE__
+  if(mutex->__align == _PTHREAD_MUTEX_SIG_init)
```
Comment (fbrausse) [inline src/c2goto/library/pthread_lib.c]: Shouldn't that be `mutex->sig`?

**Comment by fbrausse** [comment]
I am curious, do any of these 2 hold according to ESBMC on OS X? (I don't have that accessible here)
```c
#include <pthread.h>
#include <assert.h>

int main() {
  assert(sizeof(pthread_mutex_t) == 40);
  assert(__SIZEOF_PTHREAD_MUTEX_T == 40);
}
```
Because for the internal operational model, it does hold (when `__WORDSIZE == 64`) - independent of the host's libc:
https://github.com/esbmc/esbmc/blob/8e1a0b01e94e173fc47a37b21b44f4ed02f2c413/src/c2goto/headers/pthreadtypes.hs#L29-L31 Maybe we should add the `__APPLE__` definition of pthread_mutex_t to the operational model as well (similar to [pthread_cond_t](https://github.com/esbmc/esbmc/blob/8e1a0b01e94e173fc47a37b21b44f4ed02f2c413/src/c2goto/headers/pthreadtypes.hs#L123))?

**Comment by lucasccordeiro** [comment]
> I am curious, do any of these 2 hold according to ESBMC on OS X? (I don't have that accessible here)

We get a parsing error:

```
esbmc f.c
ESBMC version 6.8.0 64-bit x86_64 macos
Target: 64-bit little-endian x86_64-unknown-macos with esbmc libc
file f.c: Parsing
f.c:6:10: error: use of undeclared identifier '__SIZEOF_PTHREAD_MUTEX_T'
  assert(__SIZEOF_PTHREAD_MUTEX_T == 40);
         ^
PARSING ERROR
```

File: src/c2goto/library/pthread_lib.c (hunk 1 above)
Comment (lucasccordeiro) [reply src/c2goto/library/pthread_lib.c]: We would need to change our implementation of the pthread_mutex_t to consider this comment.

See https://github.com/esbmc/esbmc/blob/master/src/c2goto/headers/pthreadtypes.hs#L112.

**Comment by lucasccordeiro** [comment]
@fbrausse: As we discussed, I have just created this issue https://github.com/esbmc/esbmc/issues/600 to remind us to go back to this topic and discuss what to do with our pthreadtypes.hs.

File: src/c2goto/library/pthread_lib.c [hunk 2]
```diff
@@ -291,10 +295,25 @@ __ESBMC_HIDE:;
...
+int pthread_mutex_initializer(pthread_mutex_t *mutex)
+{
+  // check whether this mutex has been initialized via
+  // PTHREAD_MUTEX_INITIALIZER
+#ifdef __APPLE__
+  if(mutex->__data.__lock == _PTHREAD_MUTEX_SIG_init)
+    pthread_mutex_init(mutex, NULL);
+#else
+  if(mutex->__data.__lock)
+    pthread_mutex_init(mutex, NULL);
```
Comment (fbrausse) [inline src/c2goto/library/pthread_lib.c:310]: Won't this reset the "locked" state of an already-locked mutex?

Comment (lucasccordeiro) [reply src/c2goto/library/pthread_lib.c:310]: Fixed.

**Review by fbrausse — APPROVED** [review]
LGTM

**Review by rafaelsamenezes — APPROVED** [review]

**Review by shmarovfedor — APPROVED** [review]

**Review by kunjsong01 — APPROVED** [review]

**Pull request merged**